Firestore/core/src/firebase/firestore/local/leveldb_transaction.cc

```diff
@@ -0,0 +1,196 @@
+namespace firestore {
+namespace local {
+
+LevelDBTransaction::LevelDBTransaction(std::shared_ptr<DB> db,
+                                       const ReadOptions& readOptions,
+                                       const WriteOptions& writeOptions)
+    : db_(db), readOptions_(readOptions), writeOptions_(writeOptions) {
+}
+
+void LevelDBTransaction::Put(const std::string& key, const std::string& value) {
+  mutations_[key] = value;
+  deletions_.erase(key);
+  version_++;
+}
+
+LevelDBTransaction::Iterator::Iterator(LevelDBTransaction* txn)
+    : ldb_iter_(txn->db_->NewIterator(txn->readOptions_)),
+      txn_(txn),
+      last_version_(txn->version_),
+      is_valid_(false),  // Iterator doesn't really point to anything yet, so is
+                         // invalid
+      mutations_iter_(
+          std::make_unique<Mutations::iterator>(txn->mutations_.begin())) {
+}
+
+void LevelDBTransaction::Iterator::UpdateCurrent() {
+  bool mutation_is_valid = *mutations_iter_ != txn_->mutations_.end();
+  is_valid_ = mutation_is_valid || ldb_iter_->Valid();
+
+  if (is_valid_) {
+    if (!mutation_is_valid) {
+      is_mutation_ = false;
+    } else if (!ldb_iter_->Valid()) {
+      is_mutation_ = true;
+    } else {
+      // both are valid
+      const std::string mutation_key = (*mutations_iter_)->first;
+      const std::string ldb_key = ldb_iter_->key().ToString();
+      is_mutation_ = mutation_key <= ldb_key;
+    }
+    if (is_mutation_) {
+      current_.first = (*mutations_iter_)->first;
+      current_.second = (*mutations_iter_)->second;
+    } else {
+      current_.first = ldb_iter_->key().ToString();
+      current_.second = ldb_iter_->value().ToString();
+    }
+  }
+}
+
+void LevelDBTransaction::Iterator::Seek(const std::string& key) {
+  ldb_iter_->Seek(key);
+  for (; ldb_iter_->Valid() &&
+         txn_->deletions_.find(ldb_iter_->key().ToString()) !=
+             txn_->deletions_.end();
+       ldb_iter_->Next()) {
+  }
+  mutations_iter_.reset();
+  mutations_iter_ =
+      std::make_unique<Mutations::iterator>(txn_->mutations_.begin());
+  for (; (*mutations_iter_) != txn_->mutations_.end() &&
+         (*mutations_iter_)->first < key;
+       ++(*mutations_iter_)) {
+  }
+  UpdateCurrent();
+  last_version_ = txn_->version_;
+}
+
+std::string LevelDBTransaction::Iterator::key() {
+  FIREBASE_ASSERT_MESSAGE(Valid(), "key() called on invalid iterator");
+  return current_.first;
+}
+
+bool LevelDBTransaction::Iterator::key_starts_with(const std::string& prefix) {
+  std::string current_key = key();
+  return std::mismatch(prefix.begin(), prefix.end(), current_key.begin())
+             .first == prefix.end();
+}
+
+Slice LevelDBTransaction::Iterator::value() {
+  FIREBASE_ASSERT_MESSAGE(Valid(), "value() called on invalid iterator");
+  return current_.second;
+}
+
+bool LevelDBTransaction::Iterator::SyncToTransaction() {
+  if (last_version_ < txn_->version_) {
+    std::string current_key = current_.first;
```

Comment:
This is making a copy. Use `const std::string& current_key` to avoid it.

Reply:
The `Seek` call in the next line updates `current_`, so I believe I need to make a copy so that I can do the comparison below. I've added a comment.